Make Context to be Device-agnostic Step by Step (1/N)

fffrog opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #137580
* #137578
* #136526
* __->__ #136519

----

- make init to be device-agnostic and move it to AcceleratorHooksInterface
- refactoring context related to device initialization

fffrog committed:
Make Context to be Device-agnostic

- make init to be device-agnostic and move it to AcceleratorHooksInterface
- refactoring context related to device initialization

[ghstack-poisoned]

fffrog committed:
Update on "Make Context to be Device-agnostic Step by Step (1/N)"


- make init to be device-agnostic and move it to AcceleratorHooksInterface
- refactoring context related to device initialization

[ghstack-poisoned]

fffrog committed:
Update on "Make Context to be Device-agnostic Step by Step (1/N)"


- make init to be device-agnostic and move it to AcceleratorHooksInterface
- refactoring context related to device initialization

[ghstack-poisoned]

fffrog committed:
Update on "Make Context to be Device-agnostic Step by Step (1/N)"


- make init to be device-agnostic and move it to AcceleratorHooksInterface
- refactoring context related to device initialization

[ghstack-poisoned]

fffrog committed:
Update on "Make Context to be Device-agnostic Step by Step (1/N)"


- make init to be device-agnostic and move it to AcceleratorHooksInterface
- refactoring context related to device initialization

[ghstack-poisoned]

ezyang commented:
@pytorchbot merge

FFFrog commented:
@ezyang Thanks for your approval, can I ignore the ci failure of xpu and merge the commit into truck? I will submit a new PR in another [repo](https://github.com/intel/torch-xpu-ops) to fix xpu ci later

ezyang commented:
I don't know what xpu greenness policy is. But when we did analogous things for xla, we simultaneously updated the third party pin so that pytorch was always kept green. I suggest doing that too...

FFFrog commented:
> I don't know what xpu greenness policy is. But when we did analogous things for xla, we simultaneously updated the third party pin so that pytorch was always kept green. I suggest doing that too...

I got it, thank you.

FFFrog commented:
@EikanWang, @jgong5, sorry to bother you, could you please take a look for me? Handling xpu ci workflow failure is a little tricky for me.

EikanWang reviewed (approved):
LGTM

FFFrog commented:
> @FFFrog , thanks for the PR submission. We are running the CI. But it seems like the torch-xpu-ops will be broken due to the missed `lazyInitDevice`. We will have an ad-hoc validation to check if there are any failures.

Thank you for your comments.

@EikanWang  @guangyey , what I want to add is that do I need to update [pin commit](https://github.com/pytorch/pytorch/blob/main/third_party/xpu.txt) for xpu when this [pr](https://github.com/intel/torch-xpu-ops/pull/934) is merge later? Or there is some mechanism that can be done automatically.

guangyey commented:
> @EikanWang @guangyey , what I want to add is that do I need to update [pin commit](https://github.com/pytorch/pytorch/blob/main/third_party/xpu.txt) for xpu when this [pr](https://github.com/intel/torch-xpu-ops/pull/934) is merge later? Or there is some mechanism that can be done automatically.

Please help update xpu.txt directly once this [pr](https://github.com/intel/torch-xpu-ops/pull/934) is merged. Thanks for your help.

fffrog committed:
Update on "Make Context to be Device-agnostic Step by Step (1/N)"


- make init to be device-agnostic and move it to AcceleratorHooksInterface
- refactoring context related to device initialization

[ghstack-poisoned]

FFFrog commented:
> > @EikanWang @guangyey , what I want to add is that do I need to update [pin commit](https://github.com/pytorch/pytorch/blob/main/third_party/xpu.txt) for xpu when this [pr](https://github.com/intel/torch-xpu-ops/pull/934) is merge later? Or there is some mechanism that can be done automatically.
> 
> Please help update xpu.txt directly once this [pr](https://github.com/intel/torch-xpu-ops/pull/934) is merged. Thanks for your help.

Thank you :D

FFFrog commented:
@pytorchbot merge

ezyang commented:
@pytorchbot merge -i

cyyever commented:
@pytorchbot rebase

fffrog committed:
Update

[ghstack-poisoned]

fffrog committed:
Update on "Make Context to be Device-agnostic Step by Step (1/N)"


- make init to be device-agnostic and move it to AcceleratorHooksInterface
- refactoring context related to device initialization

[ghstack-poisoned]

cyyever commented:
@pytorchbot merge -i

FFFrog commented:
@pytorchbot merge -i

etaf commented:
@EikanWang  @FFFrog  The following failed jobs have not been resolved but PR merged, which break current XPU CI.  I'll resolve this error in PR #137585. But suggest pass XPU CI before merge PR to save triage time.
[xpu / linux-jammy-xpu-py3.9 / test (default, 3, 4, linux.idc.xpu)](https://hud.pytorch.org/pr/pytorch/pytorch/136519#31209516897) ([gh](https://github.com/pytorch/pytorch/actions/runs/11226861852/job/31209516897))
inductor/test_torchinductor_dynamic_shapes.py::DynamicShapesGPUTests::test_scatter_add3_dynamic_shapes_xpu
[xpu / linux-jammy-xpu-py3.9 / test (default, 4, 4, linux.idc.xpu)](https://hud.pytorch.org/pr/pytorch/pytorch/136519#31209517023) ([gh](https://github.com/pytorch/pytorch/actions/runs/11226861852/job/31209517023))
inductor/test_torchinductor.py::GPUTests::test_scatter_add3_xpu

jovianjaison commented:
@pytorchbot revert -m "this pr is causing errors internally" -c ghfirst
error: class "at::Context" has no member "lazyInitCUDA"

FFFrog commented:
> #137585

Sorry for the late reply.
I will update this pr once [the pr](https://github.com/pytorch/pytorch/pull/137585) is merged.

fffrog committed:
Update

[ghstack-poisoned]

EikanWang commented on third_party/xpu.txt:
We will update torch-xpu-ops commit pin first and then you can drop this change. @FFFrog

FFFrog replied:
Ok, have updated.

fffrog committed:
Update

[ghstack-poisoned]

ezyang commented:
@FFFrog so specifically the problem is there is some Caffe2 code internally that calls lazyInitCUDA. There are a few ways to fix this but I think adding aliases for the old method names is the easiest.

fffrog committed:
Update

[ghstack-poisoned]

FFFrog commented:
> @FFFrog so specifically the problem is there is some Caffe2 code internally that calls lazyInitCUDA. There are a few ways to fix this but I think adding aliases for the old method names is the easiest.

Thank you, I have updated the pr, please refer to the latest [commit](https://github.com/pytorch/pytorch/pull/136519/commits/cd098f38cc72662993a424f5df91a8a3486e05c9) to check what changed :D

ezyang commented:
@pytorchbot merge

fffrog committed:
Update

[ghstack-poisoned]

cyyever commented:
@pytorchbot merge

cyyever commented:
@pytorchbot merge -f "Unrelated failures"

cyyever commented:
@pytorchbot merge

FFFrog commented:
@pytorchbot merge

cyyever commented:
@pytorchmergebot -f "1 pending job"

cyyever commented:
@pytorchmergebot merge -f "1 pending job"

clee2000 commented:
@pytorchbot revert -m "breaking internal tests related to MITA, @ezyang has a forward fix?" -c ghfirst

ezyang commented:
I'm going to help, it's just one small problem now

ezyang commented on aten/src/ATen/detail/MTIAHooksInterface.h:
```diff
@@ -109,6 +109,11 @@ struct TORCH_API MTIAHooksInterface : AcceleratorHooksInterface {
+  // Perserved for BC
+  virtual void initMTIA() const {
+    return;
+  }
```
I don't think this should exist

ezyang commented:
I'm going to attempt to fix this directly, but a way to fix this from first principles in OSS, is to *not* rename any of the virtual methods on the hooks classes, and have a new implementation of init() in the subclass interfaces we let people subclass from that call the old method by default. What is done in this PR is cleaner but it's BC breaking and concretely what I have to do is rename initMTIA to init internally.

FFFrog commented:
My principle is to make Context.h and DeviceInterface related files clearer and easier to understand, but the changes in the current pr do introduce BC breaking, which should not be the case for a widely used project like PyTorch.

> a way to fix this from first principles in OSS, is to _not_ rename any of the virtual methods on the hooks classes

I agree with your point of view, so I updated the PR and made the following changes:
- For the in-tree backend, the implementation of device initialization is moved to the newly added **init()**, and the call of the old interface **initCUDA()** will be forwarded to init, while ensuring the code logic is as clear and usable as possible, and BC is also guaranteed
- For the out-of-tree backend, in order to ensure BC as much as possible, no changes are made to **initMTIA()**, but it is forwarded to **initMITA()** in the new interface **init()**, and a new deprecation prompt message is added

fffrog committed:
Update

[ghstack-poisoned]

fffrog committed:
Update

[ghstack-poisoned]

ezyang commented:
It turns out that the BC fix is pretty easy. I'm going to land my own copy as I need to also land an fbcode change

ezyang commented:
Internal diff with the BC update fixes https://www.internalfb.com/diff/D64471142

FFFrog commented:
@ezyang I will close this PR because the [PR](https://github.com/pytorch/pytorch/pull/138155) has been merged.
